daml-lf/spec/contract-id.rst

```diff
@@ -0,0 +1,245 @@
+contract contents include the contract instance, the template ID, the
+stakeholders / signatories / maintainers, and the contract key.
+
+**Freshness**: It is computationally infeasible to find contracts C1 and
+C2 with the same contract ID such that one of the following holds:
+
+* The contracts have different stakeholders.
+* The contracts are created by different transactions.
+* The contracts are created by different subactions of the same
+  transaction forest where the seeds are pairwise different.
+
+**Distinctness**: The ledger can enforce that the seeds are pairwise
+different.
+
+Contract Identifiers
+^^^^^^^^^^^^^^^^^^^^
+
+A *contract identifier* (or contract ID for short) is a sequence of
+bytes defined as follows ::
+
+  ContractID := verisionPrefix ∥ discriminator ∥ suffix  
+
+where
+
+* ``∥`` is the concatenation operation; 
+* ``verisionPrefix`` is 1 byte (equal to 0) used to versioned the
+  contract ID scheme;
+* ``discriminator`` is a sequence of 32 bytes. It is like a random
+  UUID, but generated from an initial seed (called *submission seed*)
+  via a fixed `derivation scheme <Allocation scheme for
+  discriminators_>`_ that supports validation (see below).
+* ``suffix`` is a sequence of 0 to 95 bytes used to enforce global
+  uniqueness of the contract ID in a distributed ledger.
+
+Discriminator freshness
+-----------------------
+
+In a transaction we distinct two kinds of contract IDs:
+
+* The *local* contract IDs are the IDs of the contracts created by the
+  transaction.
+
+*  The *global* contract IDs are the contract IDs that:
+   
+   * appears in the commands that produced the transaction. This
+     includes the IDs of the exercised contract, together with all the
+     IDs referenced in the arguments of the create and exercise
+     commands;
+   * that are fetched or looked up by key;
+   * are referenced in the input contracts.
+
+Note that local contract IDs correspond to the ID of output contracts
+together with those contracts that have been created and archived in
+the transaction. On the other hand, global contract IDs do not
+reference only IDs of some contracts that have been previously
+persisted on the ledger, but also any arbitrary contract IDs that the
+submitter have referenced in its submission.
+
+The so-called *discriminator freshness condition* holds if the
+discriminators from local contract IDs are distinct from the
+discriminators from global contract IDs.
+
+This ensures that only the discriminators, not the suffix are needed
+to order the contract IDs of created contract w.r.t. other contract
+IDs. 
+
+Contract ID uniqueness
+----------------------
+
+During interpretation local contract IDs are created without suffix.
+It is the responsibility of the ledger implementation to provide for
+each local contract ID a suffix that guarantee global uniqueness of
+the whole contract ID. No other requirement (except the 95 bytes size
+limit) is assumed for those suffices.
```

Comment:
I think this does not really cover Corda's use of suffixes, namely storing additional information in the contract ID. How about the following?

```suggestion
Ledger implementations are responsible for enforcing 
uniqueness of contract IDs on the whole ledger.
This can be done by enforcing global uniqueness of the seeds or by appropriately suffixing the contract IDs.
No other requirement (except the 95 bytes size
limit) is assumed for those suffices.
```